doc/developer/design/20250717_a_small_coordinator_more_scalable_isolated_materialize.md

```diff
@@ -0,0 +1,332 @@
+   of 100 SELECTs/s, regardless of how much capacity other parts of the system
+   have for processing.
+2. Lack of use-case isolation: When an operation takes longer than expected on
+   the main loop -- due to bugs, unexpected behavior, or other reasons -- it
+   "steals" time from other operations that need time on the loop. One use-case
+   that is using a lot of Materialize will make Materialize worse for other use
+   cases.
+
+We can see the first of these when running benchmarks. We have metrics about
+how many commands the main loop is processing and how much time they're taking
+(a histogram). And so we can see that when running a SELECT throughput
+benchmark, the time spent processing messages is close to 1 second, so at the
+theoretical limit of the architecture. More on this in a detailed section
+below.
+
+We can see the second consequence in action most acutely when there are bugs in
+operations that are thought to be fast but take a long time. In such cases,
+Materialize as a whole becomes unresponsive or "sluggish" for a customer.
+Another way of seeing this in action is of course when one use-case is using
+the environment a lot, say high SELECT throughput: this will also make the
+system less responsive for other use cases/users of the system.
+
+Both of these _might_ not obviously be problematic today, but we think we will
+have customers soon for which our scalability limits become a blocker for
+expansion and we do see a trickle of bugs where the lack of isolation makes it
+so that Materialize becomes unresponsive for customers, for multiple seconds at
+a time. This can shake confidence in Materialize, which would also be a blocker
+for adoption or expansion. There is some recency bias to thinking how urgent
+these bugs are, but here's the latest example, where we had to disable a
+feature because it can block the coordinator main loop for 10s of seconds or
+more: https://github.com/MaterializeInc/materialize/pull/33070.
+
+## Success Criteria
+
+We want to address both problems mentioned above, but scalability is easier to
+quantify. Isolation is covered by the qualitative goals.
+
+- When running throughput benchmarks, the metric showing time spent processing
+  messages on the main loop must be nowhere near 1 second. The main loop must
+  not be the bottleneck.
+- The Coordinator is so small -- in terms of the types of commands and the
+  complexity of each command -- that it is feasible to audit all commands and
+  conclude that nothing can block the main loop unreasonably long.
+
+## Out of Scope
+
+- We want to do this work within the current singleton `environmentd` process.
+  That is, we don't yet address horizontal scalability of the query/control
+  layer.
+- We don't want to improve throughput benchmark numbers, only remove the
+  Coordinator as a bottleneck. Our work might increase throughput numbers or
+  reveal similar bottlenecks in other parts of the system.
```

Comment:
I think moving the current work of peek sequencing to the frontend could increase QPS even without horizontal scaling, because:
- Doing most of the peek sequencing work on the per-session frontend task would allow us to at least vertically scale envd inside a single machine when a particular user wants a bit more QPS.
- I imagine that the current staged execution has some overhead, which would disappear if we had straight-line code instead.